src/main/java/com/amazon/opendistroforelasticsearch/knn/index/KNNWeight.java

```diff
@@ -74,57 +74,66 @@ public Scorer scorer(LeafReaderContext context) throws IOException {
-                                               ? KNNCodecUtil.HNSW_COMPOUND_EXTENSION : KNNCodecUtil.HNSW_EXTENSION;
-            String hnswSuffix = knnQuery.getField() + hnswFileExtension;
-            List<String> hnswFiles = reader.getSegmentInfo().files().stream()
-                    .filter(fileName -> fileName.endsWith(hnswSuffix))
+            FieldInfo fieldInfo = reader.getFieldInfos().fieldInfo(knnQuery.getField());
+
+            if (fieldInfo == null) {
+                logger.debug("[KNN] Field info not found for {}:{}", knnQuery.getField(),
+                        reader.getSegmentName());
+                return null;
+            }
+
+            KNNEngine knnEngine = KNNEngine.getEngine(fieldInfo.getAttribute(KNNConstants.KNN_ENGINE));
+            SpaceType spaceType = SpaceType.getSpace(fieldInfo.getAttribute(KNNConstants.SPACE_TYPE));
+            logger.debug("[KNN] knnEngine for " + knnQuery.getField() + ": " + knnEngine.getName());
+
+            String fileExtension = reader.getSegmentInfo().info.getUseCompoundFile()
+                                               ? knnEngine.getCompoundExtension() : knnEngine.getExtension();
+            String suffix = knnQuery.getField() + fileExtension;
+            List<String> engineFiles = reader.getSegmentInfo().files().stream()
+                    .filter(fileName -> fileName.endsWith(suffix))
                     .collect(Collectors.toList());
 
-            if(hnswFiles.isEmpty()) {
-                logger.debug("[KNN] No hnsw index found for field {} for segment {}",
+            if(engineFiles.isEmpty()) {
+                logger.debug("[KNN] No engine index found for field {} for segment {}",
                         knnQuery.getField(), reader.getSegmentName());
                 return null;
             }
 
             FieldInfo queryFieldInfo = reader.getFieldInfos().fieldInfo(knnQuery.getField());
             Map<String, String> fieldAttributes = queryFieldInfo.attributes();
 
-            /**
-             * in the name of the file. As of now we have one version 2.0.11 with NMSLIB and 1.6.5 with FAISS
-             * So deferring this to future releases
-             */
-
-            Path indexPath = PathUtils.get(directory, hnswFiles.get(0));
+            Path indexPath = PathUtils.get(directory, engineFiles.get(0));
             final KNNQueryResult[] results;
+            final KNNIndex index = knnIndexCache.getIndex(indexPath.toString(), knnQuery.getIndexName(), spaceType);
 
-            final KNNIndex index = knnIndexCache.getIndex(indexPath.toString(), knnQuery.getIndexName());
-            //we check instance of KNNINdex
-            if (fieldAttributes.containsValue(KNNEngine.NMSLIB.getKnnEngineName())) {
-                assert(index instanceof KNNNmsLibIndex);
-                results = index.queryIndex(
+            if ((fieldAttributes.containsValue(KNNEngine.NMSLIB.getName()) && index instanceof KNNNmsLibIndex)
```

Comment:
flip if condition to throw exception if it is not matched.
In general: if there is an exception thrown in else, you can flip if to have that part of inside if statement and everything else outside if statement. This will reduce nested level.

Reply:
that makes sense. Will fix.